### PR TITLE
Fix inconsistent maintenance "reason" key formatting

### LINKED ISSUE #2565: Maintenance reason key formatting varies between helix-rest and helix-ui
### Describe the bug
If you enable maintenance mode using a curl command against helix-rest and pass in maintenance reason in this way:
 -d "maint mode set with curl command"

The reason key will be stored in the znode in UPPERCASE formatting:
{ "REASON" : "maint mode set with curl", ... }

However, if you enable maintenance mode using helix-ui, the reason key will be stored in the znode in lowercase formatting:
{ "reason" : "maintenance mode set through UI", ... }

### To Reproduce
Create maintenance node in helix-rest and in helix-ui and observe the difference in the maintenance znode's reason key

### Expected behavior
They should have identical formatting, my opinion is that UPPERCASE formatting is intended. Hashkeys in java are case sensitive so functionally these are two separate keys when they should be identical

### Additional context
https://github.com/apache/helix/blob/4e67fdb9bbfe3124287026563e87c08f5c7c7182/helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java#L316C23-L316C23
^function that handles this request

I believe this discrepancy occurs because when you use enable it via a curl command, the data is treated as a string and the key's text is set to UPPPERCASE. 
When using helix-ui, I believe helix-rest receives it as a JSON Object of {"reason": "maintenance mode set through UI"} and and then passes it to a later function as a customsFieldMap rather than as a string. The customsFieldMap is then iterated over and the lower case "reason" key is used when setting the maintenance znode's fields

https://github.com/apache/helix/blob/4e67fdb9bbfe3124287026563e87c08f5c7c7182/helix-front/src/app/cluster/shared/cluster.service.ts#L44
^ Here is the typescript code that makes the rest request

### PR DESCRIPTION
**Pull request by GrantPSpencer**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Fixes #2565 
Formatting of the "reason" key could vary depending on how data was sent to helix-rest. This change makes it so helix-rest always checks for a "reason" key and then consistently formats it in UPPERCASE

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:
Checks the KV mapping passed to the helix-rest endpoint to determine if there a "reason" key present. We check for a case-insensitive key then set it to content, which then gets set as the value to the "REASON" (uppercase) key down the line. 

More details can be seen in the linked issue, but we need to enforce this formatting because when this endpoint receives a non-JSON string, it sets it as the value to the "REASON" key. If the endpoint receivers a json string (KV map), then it will just use whatever key-value pairs are passed to it and does not enforce UPPERCASE formatting for the reason key. This change ensures that if a reason key is present, it will always appear as "REASON" . This is important as Java maps are case sensitive and keys of different casing are functionally two different keys and do not allow customers to easily programmatically access the reason key

### Tests

- [ ] The following tests are written for this issue:
Updated testEnableDisableMaintenanceMode() in testClusterAccessor.java to assert KV map of {"reason": "Test reason"} gets stored as {"REASON": "Test reason"}

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 1]
```diff
@@ -321,8 +322,13 @@ public Response updateCluster(@PathParam("clusterId") String clusterId,
           customFieldsMap =
               OBJECT_MAPPER.readValue(content, new TypeReference<HashMap<String, String>>() {
               });
-          // content is given as a KV mapping. Nullify content
+          // content is given as a KV mapping. Nullify content unless (case-insensitive) reason key present in map
           content = null;
+          for (Map.Entry<String, String> entry : customFieldsMap.entrySet()) {
+            if ("reason".equals(entry.getKey().toLowerCase(Locale.ENGLISH))) {
```
Comment (junkaixue) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java:328]: I think Locale.ENGLISH is not needed, right? We are not checking semantics or not.

If the key does not match, it is not right then.

Comment (GrantPSpencer) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java]: Habit from old team. I've updated to use equalsIgnoreCase:

if ("reason".**equalsIgnoreCase**(entry.getKey())) {

**Review by xyuanlu — APPROVED** [review]

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java [hunk 2]
```diff
@@ -321,8 +321,13 @@ public Response updateCluster(@PathParam("clusterId") String clusterId,
           customFieldsMap =
               OBJECT_MAPPER.readValue(content, new TypeReference<HashMap<String, String>>() {
               });
-          // content is given as a KV mapping. Nullify content
+          // content is given as a KV mapping. Nullify content unless (case-insensitive) reason key present in map
           content = null;
+          for (Map.Entry<String, String> entry : customFieldsMap.entrySet()) {
```
Comment (desaikomal) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java:326]: just wondering: is  the fast way to look if the key exists? but if  we write as "reason" vs/vs "ReAsoN" then that approach won't work.

**Review by desaikomal — APPROVED** [review]
change looks good

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java (hunk 2 above)
Comment (GrantPSpencer) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java:326]: As far as I know this is the best solution to find a key case insensitive key without making changes to the map data structure, or using something a treemap like this:
TreeMap<>(String.CASE_INSENSITIVE_ORDER);

Fortunately the map should be fairly small from what I've seen, 3-10 entries, so this inefficiency shouldn't have an impact on overall performance of the call

**Comment by GrantPSpencer** [comment]
Pull request approved by @xyuanlu
Commit message: Fix inconsistent map key formatting in helix-rest maintenance API

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java (hunk 2 above)
Comment (desaikomal) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java:326]: good to know.. thanks for working on this.

**Pull request merged**